test/parallel_api/numeric/numeric.ops/in_place_scan.pass.cpp

```diff
@@ -19,64 +19,479 @@
+        }
+    }
+
+    // specialization for non-random_access iterators
+    template <typename Policy, typename Iterator1, typename Iterator2, typename Size>
+    typename ::std::enable_if<!is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value, void>::type
+    operator()(Policy&& exec,
+               Iterator1 keys_first, Iterator1 keys_last,
+               Iterator2 vals_first, Iterator2 vals_last,
+               Size n)
+    {
+    }
+};
+
+DEFINE_TEST_1(test_inclusive_scan_inplace, BinaryOperation)
+{
+    DEFINE_TEST_CONSTRUCTOR(test_inclusive_scan_inplace)
+
+    template <typename TData, typename Iterator1, typename Size,
+              typename BinaryOperationCheck = oneapi::dpl::__internal::__pstl_plus>
+    void check_values(const std::vector<TData>& source_host_keys_state, Iterator1 host_keys, Size n,
+                      BinaryOperationCheck op = BinaryOperationCheck())
+    {
+        std::vector<TData> expected(n);
+
+        inclusive_scan_serial(source_host_keys_state.begin(), source_host_keys_state.end(), expected.begin(), op);
+
+        EXPECT_EQ_N(expected.cbegin(), host_keys, n, "wrong effect from inplace inclusive_scan");
+    }
+
+#if TEST_DPCPP_BACKEND_PRESENT
+    // specialization for hetero policy
+    template <typename Policy, typename Iterator1, typename Size>
+    typename ::std::enable_if<
+        oneapi::dpl::__internal::__is_hetero_execution_policy<typename ::std::decay<Policy>::type>::value &&
+            is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value,
+        void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last,
+               Size n)
+    {
+        using KeyT = typename ::std::iterator_traits<Iterator1>::value_type;
+
+        TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);
+
+        // copy source host keys state into temp buf
+        initialize_data(host_keys.get(), n);
+        const std::vector<KeyT> source_host_keys_state(host_keys.get(), host_keys.get() + n);
+
+        // Call inplace inclusive_scan algorithm with hetero policy
+        {
+            update_data(host_keys);
+
+            auto new_policy = make_new_policy<new_kernel_name<Policy, 0>>(exec);
+            oneapi::dpl::inclusive_scan(new_policy, keys_first, keys_last, keys_first);
+
+            retrieve_data(host_keys);
+            check_values(source_host_keys_state, host_keys.get(), n);
+        }
+
+        // Call inplace inclusive_scan algorithm with hetero policy and user operation
+        {
+            initialize_data(host_keys.get(), n);
+            update_data(host_keys);
+
+            auto new_policy1 = make_new_policy<new_kernel_name<Policy, 1>>(exec);
+            oneapi::dpl::inclusive_scan(new_policy1, keys_first, keys_last, keys_first, BinaryOperation());
+
+            retrieve_data(host_keys);
+            check_values(source_host_keys_state, host_keys.get(), n, BinaryOperation());
```

Comment:
Here we calculate implicitly the expected values the second time.. The first one have on line 212.
It would be better to do allocation and expected calculation one time 
```
std::vector<TData> expected(n);
inclusive_scan_serial
```
above line 204.

Reply:
Hope I understood right what you mean. Fixed.